### PR TITLE
Add 2d domain decomposition to the inline post

### PR DESCRIPTION
**Pull request by JesseMeng-NOAA**
## Description

Resubmit an update from previous PR, after many updates in upp and other fv3atm submodules.

The community upp is currently decomposed on latitudes (y-direction) only. As the ufs-weather-model is making 2D decomposition in both latitudes and longitudes (Y and X), one of the upp refactoring tasks is to support ufs-weather-model inline post 2D decomposition functionality.

A ufs-weather-model PR was created for this test with inline post turned on in the control_2dwrtdecomp and regional_control_2dwrtdecomp tests,
https://github.com/ufs-community/ufs-weather-model/pull/1211

The upp/post_2d_decomp branch can be found at
https://github.com/WenMeng-NOAA/UPP/tree/post_2d_decomp

(Instructions: this, and all subsequent sections of text should be removed and filled in as appropriate.)
Provide a detailed description of what this PR does.  
What bug does it fix, or what feature does it add?

Several io/ subroutines are modified to call upp/post_2d_decomp functions
   
Is a change of answers expected from this PR?  

This PR does not change the output.

### Issue(s) addressed

https://github.com/ufs-community/ufs-weather-model/issues/1078

Link the issues to be closed with this PR, whether in this repository, or in another repository.
(Remember, issues should always be created before starting work on a PR branch!)

## Testing

How were these changes tested?

1. Build the BASE with ufs-weather-model/develop
    run control fv3 and regional_control fv3  

2. Checkout the branch for this PR
    https://github.com/JesseMeng-NOAA/fv3atm/tree/feature/upp_unify_inline_2d
    run control_2dwrtdecomp and regional_control_2dwrtdecomp

3. Compare the results to confirm identical outputs are reproduced.

What compilers / HPCs was it tested with?  
Are the changes covered by regression tests? (If not, why? Do new tests need to be added?)  
Have the ufs-weather-model regression test been run? On what platform?  
- Will the code updates change regression test baseline? If yes, why? Please show the baseline directory below.
- Please commit the regression test log files in your ufs-weather-model branch

## Dependencies

https://github.com/WenMeng-NOAA/UPP/tree/post_2d_decomp

### CODE REVIEW

File: io/post_fv3.F90 [hunk 1]
```diff
@@ -852,8 +862,9 @@ subroutine set_postvars_fv3(wrt_int_state,mpicomp,setvar_atmfile,   &
                 line=__LINE__, file=__FILE__)) return  ! bail out
 !           print *,'in post_lam, get land field value,fillvalue=',fillvalue
 
-          ista = lbound(arrayr42d,1)
-          iend = ubound(arrayr42d,1)
+!JESSE 20220222 keep consistent ista/iend but keep the footprint of lbound/ubound setting
+!          ista = lbound(arrayr42d,1)
+!          iend = ubound(arrayr42d,1)
```
Comment (DusanJovic-NOAA) [inline io/post_fv3.F90]: Please remove commented, unused code.

File: io/post_fv3.F90 [hunk 2]
```diff
@@ -886,8 +897,9 @@ subroutine set_postvars_fv3(wrt_int_state,mpicomp,setvar_atmfile,   &
                 line=__LINE__, file=__FILE__)) return  ! bail out
 !           if(mype==0) print *,'in post_lam, get icec  field value,fillvalue=',fillvalue
 
-          ista = lbound(arrayr42d,1)
-          iend = ubound(arrayr42d,1)
+!JESSE 20220222 keep consistent ista/iend but keep the footprint of lbound/ubound setting
+!          ista = lbound(arrayr42d,1)
+!          iend = ubound(arrayr42d,1)
```
Comment (DusanJovic-NOAA) [inline io/post_fv3.F90]: Please remove commented, unused code

File: io/post_nems_routines.F90 [hunk 3]
```diff
@@ -100,22 +126,74 @@ subroutine post_alctvars(imi,jmi,lmi,mype,nwtlpes,lead_write, mpicomp,  &
...
+      allocate(jexa(0:num_procs-1) )
+
       do i = 1, num_procs
-       icnt(i-1) = (jtegrp(i)-jtsgrp(i)+1)*im
-       idsp(i-1) = (jtsgrp(i)-1)*im
+!       icnt(i-1) = (jtegrp(i)-jtsgrp(i)+1)*im
+!       idsp(i-1) = (jtsgrp(i)-1)*im
 !       if ( mype .eq. lead_write ) then
 !           print *, ' i, icnt(i),idsp(i) = ',i-1,icnt(i-1),idsp(i-1)
 !       end if
```
Comment (DusanJovic-NOAA) [inline io/post_nems_routines.F90]: Please remove commented, unused code

File: io/post_nems_routines.F90 [hunk 4]
```diff
@@ -78,14 +86,32 @@ subroutine post_alctvars(imi,jmi,lmi,mype,nwtlpes,lead_write, mpicomp,  &
...
+      if(mod(me,numx)==0)then
+        ista_m=2
+        ista_m2=3
+      end if
+      if(mod(me+1,numx)==0)then
+        iend_m=im-1
+        iend_m2=im-2
+      end if
+
+
```
Comment (junwang-noaa) [inline io/post_nems_routines.F90]: Please remove extra line

File: io/post_nems_routines.F90 [hunk 5]
```diff
@@ -100,22 +126,74 @@ subroutine post_alctvars(imi,jmi,lmi,mype,nwtlpes,lead_write, mpicomp,  &
...
       jvend_2u = min(jend + 2, jm+1 )
       if(mype==0)print *,'im=',im,'jsta_2l=',jsta_2l,'jend_2u=',jend_2u,'lm=',lm
-!
+      print 901,'GWVX mype/me=',mype,me,'im=',im,'jsta   =',jsta   ,'jend   =',jend   ,'lm=',lm
+      print 901,'GWVX mype/me=',mype,me,'im=',im,'jsta_m =',jsta_m ,'jend_m =',jend_m ,'lm=',lm
+      print 901,'GWVX mype/me=',mype,me,'im=',im,'jsta_2l=',jsta_2l,'jend_2u=',jend_2u,'lm=',lm
+      print 901,'GWVX mype/me=',mype,me,'im=',im,'ista   =',ista   ,'iend   =',iend   ,'lm=',lm
+      print 901,'GWVX mype/me=',mype,me,'im=',im,'ista_m =',ista_m ,'iend_m =',iend_m ,'lm=',lm
+      print 901,'GWVX mype/me=',mype,me,'im=',im,'ista_2l=',ista_2l,'iend_2u=',iend_2u,'lm=',lm
+  901 format(a15,2i4,4(1x,a8,i4))
```
Comment (DusanJovic-NOAA) [inline io/post_nems_routines.F90]: Remove debug prints.

File: io/post_nems_routines.F90 [hunk 6]
```diff
@@ -100,22 +126,74 @@ subroutine post_alctvars(imi,jmi,lmi,mype,nwtlpes,lead_write, mpicomp,  &
...
+      allocate(isxa(0:num_procs-1) )
+      allocate(jsxa(0:num_procs-1) )
+      allocate(iexa(0:num_procs-1) )
+      allocate(jexa(0:num_procs-1) )
+
       do i = 1, num_procs
-       icnt(i-1) = (jtegrp(i)-jtsgrp(i)+1)*im
-       idsp(i-1) = (jtsgrp(i)-1)*im
+!       icnt(i-1) = (jtegrp(i)-jtsgrp(i)+1)*im
+!       idsp(i-1) = (jtsgrp(i)-1)*im
```
Comment (junwang-noaa) [inline io/post_nems_routines.F90]: Please remove all the commented lines.

File: io/post_nems_routines.F90 [hunk 7]
```diff
@@ -100,22 +126,74 @@ subroutine post_alctvars(imi,jmi,lmi,mype,nwtlpes,lead_write, mpicomp,  &
...
+      print 901,'GWVX mype/me=',mype,me,'im=',im,'ista_m =',ista_m ,'iend_m =',iend_m ,'lm=',lm
+      print 901,'GWVX mype/me=',mype,me,'im=',im,'ista_2l=',ista_2l,'iend_2u=',iend_2u,'lm=',lm
+  901 format(a15,2i4,4(1x,a8,i4))
+!       NEW neighbors
+      ileft = me - 1
+      iright = me + 1
+       iup=MPI_PROC_NULL
+       idn=MPI_PROC_NULL
+    if(mod(me,numx) .eq. 0) print *,' LEFT POINT',mype,me
+    if(mod(me+1,numx) .eq. 0) print *,' RIGHT  POINT',mype,me
```
Comment (DusanJovic-NOAA) [inline io/post_nems_routines.F90]: Remove debug prints.

File: io/post_nems_routines.F90 [hunk 8]
```diff
@@ -100,22 +126,74 @@ subroutine post_alctvars(imi,jmi,lmi,mype,nwtlpes,lead_write, mpicomp,  &
...
+       if(jtsgrp(i)==1 .or. jtegrp(i)==jm) then
+          icnt2(i-1) = (itegrp(i)-itsgrp(i)+1)
+       else
+          icnt2(i-1) = 0
+       endif
+       idsp2(i-1)=isumm2
+       if(jtsgrp(i)==1 .or. jtegrp(i)==jm) isumm2=isumm2+(itegrp(i)-itsgrp(i)+1)
+
       enddo
+      write(6,'(a25,i4,16i8)') 'JESSE,me,icnt2,idsp2=',me,icnt2(0:num_procs-1),idsp2(0:num_procs-1)
```
Comment (junwang-noaa) [inline io/post_nems_routines.F90]: Please remove this print line

Comment (DusanJovic-NOAA) [inline io/post_nems_routines.F90]: Remove debug prints.

File: io/post_nems_routines.F90 [hunk 9]
```diff
@@ -100,22 +126,74 @@ subroutine post_alctvars(imi,jmi,lmi,mype,nwtlpes,lead_write, mpicomp,  &
...
 ! special for c-grid v
       jvend_2u = min(jend + 2, jm+1 )
       if(mype==0)print *,'im=',im,'jsta_2l=',jsta_2l,'jend_2u=',jend_2u,'lm=',lm
-!
+      print 901,'GWVX mype/me=',mype,me,'im=',im,'jsta   =',jsta   ,'jend   =',jend   ,'lm=',lm
+      print 901,'GWVX mype/me=',mype,me,'im=',im,'jsta_m =',jsta_m ,'jend_m =',jend_m ,'lm=',lm
+      print 901,'GWVX mype/me=',mype,me,'im=',im,'jsta_2l=',jsta_2l,'jend_2u=',jend_2u,'lm=',lm
+      print 901,'GWVX mype/me=',mype,me,'im=',im,'ista   =',ista   ,'iend   =',iend   ,'lm=',lm
+      print 901,'GWVX mype/me=',mype,me,'im=',im,'ista_m =',ista_m ,'iend_m =',iend_m ,'lm=',lm
+      print 901,'GWVX mype/me=',mype,me,'im=',im,'ista_2l=',ista_2l,'iend_2u=',iend_2u,'lm=',lm
```
Comment (junwang-noaa) [inline io/post_nems_routines.F90]: These print lines will be output from all the write tasks. Please comment them out.

File: io/post_nems_routines.F90 (hunk 7 above)
Comment (junwang-noaa) [inline io/post_nems_routines.F90]: Please comment out the debug lines.

File: io/post_nems_routines.F90 [hunk 10]
```diff
@@ -78,14 +86,32 @@ subroutine post_alctvars(imi,jmi,lmi,mype,nwtlpes,lead_write, mpicomp,  &
       jsta_m2 = jsta
       jend_m  = jend
       jend_m2 = jend
-      if ( mype == lead_write ) then
+!      if ( mype == lead_write ) then
```
Comment (DusanJovic-NOAA) [inline io/post_nems_routines.F90]: Please remove commented, unused code

File: io/post_nems_routines.F90 [hunk 11]
```diff
@@ -78,14 +86,32 @@ subroutine post_alctvars(imi,jmi,lmi,mype,nwtlpes,lead_write, mpicomp,  &
...
+      if ( mype<(lead_write+numx) ) then
          jsta_m  = 2
          jsta_m2 = 3
       end if
-      if ( mype == last_write_task ) then
+!      if ( mype == last_write_task ) then
```
Comment (DusanJovic-NOAA) [inline io/post_nems_routines.F90]: Please remove commented, unused code

File: io/post_fv3.F90 (hunk 1 above)
Comment (JesseMeng-NOAA) [reply io/post_fv3.F90]: Done.

File: io/post_fv3.F90 (hunk 2 above)
Comment (JesseMeng-NOAA) [reply io/post_fv3.F90]: Done.

File: io/post_nems_routines.F90 (hunk 3 above)
Comment (JesseMeng-NOAA) [reply io/post_nems_routines.F90]: Done.

File: io/post_nems_routines.F90 (hunk 4 above)
Comment (JesseMeng-NOAA) [reply io/post_nems_routines.F90]: Done.

File: io/post_nems_routines.F90 (hunk 5 above)
Comment (JesseMeng-NOAA) [reply io/post_nems_routines.F90]: Done.

File: io/post_nems_routines.F90 (hunk 6 above)
Comment (JesseMeng-NOAA) [reply io/post_nems_routines.F90]: Done.

File: io/post_nems_routines.F90 (hunk 7 above)
Comment (JesseMeng-NOAA) [reply io/post_nems_routines.F90]: Done.

File: io/post_nems_routines.F90 (hunk 8 above)
Comment (JesseMeng-NOAA) [reply io/post_nems_routines.F90]: Done.

Comment (JesseMeng-NOAA) [reply io/post_nems_routines.F90]: Done.

File: io/post_nems_routines.F90 (hunk 9 above)
Comment (JesseMeng-NOAA) [reply io/post_nems_routines.F90]: Done.

File: io/post_nems_routines.F90 (hunk 7 above)
Comment (JesseMeng-NOAA) [reply io/post_nems_routines.F90]: Done.

File: io/post_nems_routines.F90 (hunk 10 above)
Comment (JesseMeng-NOAA) [reply io/post_nems_routines.F90]: Done.

File: io/post_nems_routines.F90 (hunk 11 above)
Comment (JesseMeng-NOAA) [reply io/post_nems_routines.F90]: Done.

File: .gitmodules [hunk 12]
```diff
@@ -12,5 +12,5 @@
   branch = main
 [submodule "upp"]
   path = upp
-  url = https://github.com/NOAA-EMC/UPP
+  url = https://github.com/NOAA-EMC/UPP.git
```
Comment (DusanJovic-NOAA) [inline .gitmodules]: Remove '.git' suffix

**Comment by jkbk2004** [comment]
@DusanJovic-NOAA @junwang-noaa @JesseMeng-NOAA I think we can merge in this PR. Can you check again and leave any comment if needed?

**Review by junwang-noaa — APPROVED** [review]

**Review by DusanJovic-NOAA — APPROVED** [review]

**Review by WenMeng-NOAA — APPROVED** [review]

**Pull request merged**